ddl/column.go

```diff
@@ -693,17 +705,389 @@ func (w *worker) doModifyColumn(
+		})
+		if err != nil {
+			if errWaitReorgTimeout.Equal(err) {
+				// If timeout, we should return, check for the owner and re-wait job done.
+				return ver, nil
+			}
+			if kv.ErrKeyExists.Equal(err) || errCancelledDDLJob.Equal(err) || errCantDecodeIndex.Equal(err) {
+				logutil.BgLogger().Warn("[ddl] run modify column job failed, convert job to rollback", zap.String("job", job.String()), zap.Error(err))
+				// TODO: Do rollback.
+			}
+			if types.ErrOverflow.Equal(err) {
+				// TODO: Do rollback.
+				job.State = model.JobStateCancelled
+				return ver, errors.Trace(err)
+			}
+			// Clean up the channel of notifyCancelReorgJob. Make sure it can't affect other jobs.
+			w.reorgCtx.cleanNotifyReorgCancel()
+			return ver, errors.Trace(err)
+		}
+		// Clean up the channel of notifyCancelReorgJob. Make sure it can't affect other jobs.
+		w.reorgCtx.cleanNotifyReorgCancel()
+
+		// Remove the old column and indexes. Update the relative column name and index names.
+		tblInfo.Columns = tblInfo.Columns[:len(tblInfo.Columns)-1]
+		for _, cIdx := range changingIdxs {
+			idxName := strings.TrimPrefix(cIdx.Name.O, changingIndexPrefix)
+			for i, idx := range tblInfo.Indices {
+				if strings.EqualFold(idxName, idx.Name.L) {
+					cIdx.Name = model.NewCIStr(idxName)
+					tblInfo.Indices[i] = cIdx
+					break
+				}
+			}
+		}
+		changingCol.Name = colName
+		tblInfo.Indices = tblInfo.Indices[:len(tblInfo.Indices)-len(changingIdxs)]
+		// Adjust table column offset.
+		if err = adjustColumnInfoInModifyColumn(job, tblInfo, changingCol, oldCol, pos); err != nil {
+			// TODO: Do rollback.
+			return ver, errors.Trace(err)
+		}
+		updateChangingInfo(job, changingCol, changingIdxs, model.StatePublic)
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != changingCol.State)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// Finish this job.
+		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+		// TODO: Change column ID.
+		// asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfos: []*model.ColumnInfo{changingCol}})
+	default:
+		err = ErrInvalidDDLState.GenWithStackByArgs("column", changingCol.State)
+	}
+
+	return ver, errors.Trace(err)
+}
+
+func (w *worker) updatePhysicalTableRow(t table.PhysicalTable, oldColInfo, colInfo *model.ColumnInfo, reorgInfo *reorgInfo) error {
+	logutil.BgLogger().Info("[ddl] start to update table row", zap.String("job", reorgInfo.Job.String()), zap.String("reorgInfo", reorgInfo.String()))
+	return w.writePhysicalTableRecord(t.(table.PhysicalTable), typeUpdateColumnWorker, nil, oldColInfo, colInfo, reorgInfo)
+}
+
+// updateColumnAndIndexes handles the modify column reorganization state for a table.
+func (w *worker) updateColumnAndIndexes(t table.Table, oldCol, col *model.ColumnInfo, idxes []*model.IndexInfo, reorgInfo *reorgInfo) error {
+	// TODO: Consider rebuild ReorgInfo key to mDDLJobReorgKey_jobID_elementID(colID/idxID).
```

Comment:
@zimulala will you implements this TODO on this PR?

Reply:
No, I will handle it in the next PR.

Reply:
OK, #19303 also require this feature.